vivado/defs.bzl

```diff
@@ -618,3 +646,97 @@ def vivado_flow(name, module, module_top, part_number, xilinx_env, tags = []):
         xilinx_env = xilinx_env,
         tags = tags,
     )
+
+def _xsim_test_impl(ctx):
+    all_files, hdl_source_content, constraints_content, tcl_content = generate_file_load_tcl(ctx.attr.module)
+
+    project_dir = ctx.actions.declare_directory("{}_prj".format(ctx.label.name))
+    if (ctx.attr.with_waveform):
+        with_waveform_str = "1"
+        wave_db = ctx.actions.declare_file("{}.wdb".format(ctx.label.name))
+        wave_db_path = wave_db.path
+        outputs = [project_dir, wave_db]
+    else:
+        with_waveform_str = "0"
+        wave_db_path = ""
+        outputs = [project_dir]
+
+    substitutions = {
+        "{{PART_NUMBER}}": ctx.attr.part_number,
+        "{{HDL_SOURCE_CONTENT}}": hdl_source_content,
+        "{{TCL_CONTENT}}": tcl_content,
+        "{{CONSTRAINTS_CONTENT}}": constraints_content,
+        "{{MODULE_TOP}}": ctx.attr.module_top,
+        "{{PROJECT_DIR}}": project_dir.path,
+        "{{WITH_WAVEFORM}}": with_waveform_str,
+        "{{WAVE_DB}}": wave_db_path,
+    }
+
+    _, vivado_log, vivado_journal = run_tcl_template(
+        ctx,
+        ctx.file._xsim_test_template,
+        substitutions,
+        ctx.file.xilinx_env,
+        all_files,
+        outputs,
+    )
+
+    outputs.append(vivado_log)
+    outputs.append(vivado_journal)
+
+    log_runfiles = ctx.runfiles(files = [vivado_log])
+
+    error_parser = "if grep -q Error {}; then\n".format(vivado_log.short_path)
+    error_parser += "cat {}\n".format(vivado_log.short_path)
+    error_parser += "exit 64\n"
+    error_parser += "fi"
+
+    ctx.actions.write(
```

Comment:
We should move this error checker to a validation action.

https://bazel.build/extending/rules#validation_actions

Reply:
I hadn't seen those thanks!

I am a bit unsure about how to make a test with a validating action though. It seems like it's meant as a consistency check on building a blob or some other artifact. For a test rule, it is required to return an executable to evaluate if the test passed or failed.

To use a validating action in a test rule I would need to somehow make it generate an executable file to return in the default info. If this is possible, could you point me to an example or provide a code snippet?

Reply:
That might be an issue. I'll mark this as resolved for now